In [10]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import xgboost as xgb
from sklearn.metrics import mean_absolute_error, r2_score, mean_squared_error
from datetime import datetime

file_path = '/content/Cleaned Data (2).csv'
data = pd.read_csv(file_path)

data['Date/Time'] = pd.to_datetime(data['Date/Time'])

data['Hour'] = data['Date/Time'].dt.hour
data['Day'] = data['Date/Time'].dt.day
data['Month'] = data['Date/Time'].dt.month
data.drop(columns=['Date/Time', 'Theoretical_Power_Curve (KWh)'], inplace=True)

X = data.drop(columns=['LV ActivePower (kW)'])
y = data['LV ActivePower (kW)']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=5, random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
accuracy = 100 - (mae / np.mean(y_test) * 100)
print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'R-squared Score: {r2}')
print(f'Model Accuracy: {accuracy:.2f}%')

# User input for prediction
def predict_lv_activation():
    print("Enter values for prediction:")
    date_time_str = input("Enter Date/Time (YYYY-MM-DD HH:MM:SS): ").strip()
    date_time = datetime.strptime(date_time_str, "%Y-%m-%d %H:%M:%S")
    hour = date_time.hour
    day = date_time.day
    month = date_time.month
    wind_speed = float(input("Wind Speed (m/s): "))
    wind_direction = float(input("Wind Direction (°): "))

    user_data = np.array([[wind_speed, wind_direction, hour, day, month]])
    user_data = scaler.transform(user_data)
    prediction = model.predict(user_data)
    print(f'Predicted LV ActivePower (kW): {prediction[0]}')


predict_lv_activation()


Mean Absolute Error: 76.79123823493927
Mean Squared Error: 32198.799175621327
R-squared Score: 0.980330116313581
Model Accuracy: 95.27%
Enter values for prediction:
Enter Date/Time (YYYY-MM-DD HH:MM:SS): 2024-03-01 15:30:00
Wind Speed (m/s): 12.5
Wind Direction (°): 180
Predicted LV ActivePower (kW): 3426.082763671875


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


Enter Date/Time (YYYY-MM-DD HH:MM:SS): 2024-03-01 15:30:00  
Wind Speed (m/s): 12.5  
Wind Direction (°): 180  